core/src/connection_reuse.rs

```diff
@@ -40,77 +40,261 @@
+        // Keys of the elements in `connections` to remove afterwards.
+        let mut to_remove = Vec::new();
+        // Substream to return, if any found.
+        let mut ret_value = None;
+        let mut found_one = false;
+
+        for (addr, state) in self.connections.lock().iter_mut() {
+            let res = {
+                if let PeerState::Active {
+                    ref mut muxer,
+                    ref mut num_substreams,
+                    connection_id,
+                    client_addr,
+                    listener_id,
+                } = state
+                {
+                    if *listener_id == listener {
+                        continue;
+                    }
+                    found_one = true;
+
+                    match muxer.clone().inbound().poll() {
+                        Ok(Async::Ready(Some(inner))) => {
+                            trace!("New incoming substream from {}", client_addr);
+                            *num_substreams += 1;
+                            Ok((inner, connection_id.clone(), client_addr.clone()))
+                        }
+                        Ok(Async::Ready(None)) => {
+                            // The muxer isn't capable of opening any inbound stream anymore, so
+                            // we close the connection entirely.
+                            trace!("Removing existing connection to {} as it cannot open inbound anymore", addr);
+                            to_remove.push(addr.clone());
+                            continue;
+                        }
+                        Ok(Async::NotReady) => continue,
+                        Err(err) => {
+                            // If an error happens while opening an inbound stream, we close the
+                            // connection entirely.
+                            trace!(
+                                "Error while opening inbound substream to {}: {:?}",
+                                addr,
+                                err
+                            );
+                            to_remove.push(addr.clone());
+                            Err(err)
+                        }
+                    }
+                } else {
+                    if listener.is_none() {
+                        if let PeerState::Pending { ref mut notify, .. } = state {
+                            notify.insert(TASK_ID.with(|&t| t), task::current());
+                        }
+                    }
+                    continue;
+                }
+            };
+
+            ret_value = Some(res);
+            break;
+        }
+
+        for to_remove in to_remove {
+            self.connections.lock().remove(&to_remove);
+        }
+
+        match ret_value {
+            Some(Ok(val)) => Ok(Async::Ready(Some(val))),
+            Some(Err(err)) => Err(err),
+            None => {
+                if found_one {
+                    Ok(Async::NotReady)
```

Comment:
If all muxers return `Ok(Async::Ready(None))` we would need to register the current task for eventual notification I think.